containers/source-build/app/source_build.py

```diff
@@ -0,0 +1,465 @@
+
+            extract_blob_member(
+                digest,
+                blob_member,
+                dest_dir,
+                rename_to=os.path.basename(symlink_member),
+                work_dir=base_sources_extraction_dir,
+                log=log,
+            )
+
+            gathered = True
+
+        elif symlink_member.startswith("./extra_src_dir/"):
+            extra_src_archive = os.path.basename(symlink_member)
+            log.debug("Prepare extra source %s", extra_src_archive)
+            # one extra source archive per directory, no matter what the directory name is.
+            dest_dir = create_dir(extra_src_dir, extra_src_archive)
+            sib_dirs.extra_src_dirs.append(dest_dir)  # save this directory for bsi
+
+            extract_blob_member(
+                digest,
+                blob_member,
+                str(dest_dir),
+                rename_to=extra_src_archive,
+                work_dir=base_sources_extraction_dir,
+                log=log,
+            )
+
+            gathered = True
+
+            # FIXME: perhaps the dependency archive (the Cachi2) might be handled differently
+            run(["tar", "xvf", extra_src_archive], check=True, cwd=dest_dir)
+            os.unlink(f"{dest_dir}/{extra_src_archive}")
+
+        else:
+            log.warning("No known operation happened on layer %s", digest)
+
+    return gathered
+
+
+def gather_prefetched_sources(
+    work_dir: str, cachi2_dir: str, sib_dirs: SourceImageBuildDirectories
+) -> bool:
+    log = logging.getLogger("source-build.prefetched-sources")
+    gathered = False
+
+    # Guess if hermetic build is enabled
+    # NOTE: this guess does depend on how cachi2 runs inside prefetch-dependencies task.
+    cachi2_output_dir = f"{cachi2_dir}/output"
+
+    if not os.path.isdir(cachi2_output_dir):
+        log.info("Cannot find cachi2 output directory at %s", cachi2_output_dir)
+        return gathered
+
+    def _find_prefetch_source_archives():
+        guess_mime = filetype.guess_mime
+        for root, dirs, files in os.walk(cachi2_output_dir):
+            for filename in files:
+                mimetype = guess_mime(os.path.join(root, filename))
+                if mimetype and mimetype in ARCHIVE_MIMETYPES:
+                    yield root, filename
+
+    source_counter = itertools.count()
+    prepared_sources_dir = create_dir(work_dir, "prefetched_sources")
+    relative_to = os.path.relpath
+
+    for root, filename in _find_prefetch_source_archives():
+        src_dir = f"src-{next(source_counter)}"
+        copy_dest_dir = f"{prepared_sources_dir}/{src_dir}/{relative_to(root, cachi2_output_dir)}"
+        os.makedirs(copy_dest_dir)
+
+        src = f"{root}/{filename}"
+        dest = f"{copy_dest_dir}/{filename}"
+        log.debug("copy prefetched source %s to %s", src, dest)
+        shutil.copy(src, dest)
+        sib_dirs.extra_src_dirs.append(f"{prepared_sources_dir}/{src_dir}")
```

Comment:
So currently, this is one `extra_src_dir` for each archive in the cachi2 output, right? And, if I'm not mistaken, BSI puts each extra_src_dir in a separate layer?

This can easily result in source containers with 1000s of layers. Is that OK?

Reply:
Original idea of source containers was that because each source is in separate layer, it will be a mechanism for deduplication of data, as quay can reuse layers in multiple image. I'm not sure how this is working in practice, any byte change would result into different digest and no data deduplication is happening.

Also in OCI each layer can have annotation (and BSI should be filling this), so you should be able to find exact source by checking just index annotations, and the just fetch the source you care about.

Reply:
Ok, sounds like more layers is good then. The archive fetched by cachi2 for a specific name==version of a dependency should be byte-for-byte identical every time (if the user is verifying checksums)

Reply:
> Unless filesystem metadata matters as well (timestamp)
This is what bothers me, but there is no other way than testing it :)

Reply:
Hmm, but I ran into an issue when testing this with quay

```bash
python containers/source-build/app/source_build.py \
  --bsi /tmp/BSI/BuildSourceImage.sh \
  -s /tmp/test/quay \
  -w source-ws \
  --cachi2-artifacts-dir /tmp/test/cachi2 \
  --output-binary-image foo 1>stdout.log 2>stderr.log
```
```bash
$ grep 'too long' stderr.log
/tmp/BSI/BuildSourceImage.sh: line 1049: /tmp/work/build-definitions/source-ws/source-build/bsi_build/tmp/SrcImg.S4EPU8/driver/extra_src_dir/extra-src-011111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111.tar.json: File name too long
```
Fun fact: line 2775 in stderr.log (the one just before the error, which logs the bsi command) is 243_000 characters long :smile:

Reply:
It looks like BSI will only be able to handle less than 300 extra sources because of the naming it uses for the files (0, 01, 011, 0111, ...)

Reply:
And yes, it is just a stupid bash bug in BSI :smile: 

I'm pretty sure that the intention wasn't

```bash
counter=0
counter+=1
# now counter is "01"
```
but rather
```bash
counter=0
((counter+=1))
# now counter is "1"

Reply:
buildah source?

Reply:
I thought `((counter+=1))` is the one bsi and we want, but not sure `counter+=1` is intended.

Reply:
Maybe we can apply that as another patch in the Dockerfile? Until we drop BSI and move to `buildah source`?